src/model_theory/syntax.lean

```diff
@@ -66,13 +66,60 @@ namespace term
 
 open list
 
+/-- Indicates whether a variable is used in a given term. -/
+@[simp] def uses_var : L.term α → α → Prop
+| (var i) a := i = a
+| (func f ts) a := ∃ i, (ts i).uses_var a
+
+lemma var_uses_var (a : α) : (var a : L.term α).uses_var a :=
+by rw uses_var
+
+lemma func_uses_var {n : ℕ} {f : L.functions n} {a : α} {ts : fin n → L.term α} {i : fin n}
+  (h : (ts i).uses_var a) :
+  (func f ts).uses_var a :=
+⟨i, h⟩
+
+lemma finite_uses_var (t : L.term α) : {a | t.uses_var a}.finite :=
+begin
+  induction t with _ _ _ _ ih,
+  { simp only [uses_var, set.set_of_eq_eq_singleton', set.finite_singleton], },
+  { simp_rw [uses_var, set.set_of_exists],
+    exact set.finite_Union ih }
+end
+
+lemma finite_uses_var_left (t : L.term (α ⊕ β)) : {a | t.uses_var (sum.inl a)}.finite :=
+t.finite_uses_var.preimage (sum.inl_injective.inj_on _)
+
 /-- Relabels a term's variables along a particular function. -/
 @[simp] def relabel (g : α → β) : L.term α → L.term β
 | (var i) := var (g i)
 | (func f ts) := func f (λ i, (ts i).relabel)
 
-instance inhabited_of_var [inhabited α] : inhabited (L.term α) :=
-⟨var default⟩
+/-- Restricts a term to use only a set of the given variables. -/
+def restrict_var (s : set α) : Π (t : L.term α) (h : ∀ a, t.uses_var a → a ∈ s), L.term s
+| (var a) h := var ⟨a, h a rfl⟩
+| (func f ts) h := func f (λ i, (ts i).restrict_var (λ a ha, h a ⟨i, ha⟩))
+
+/-- A helper function for `restrict_var_left`. -/
+def restrict_var_left_aux (s : set α) :
+  (sum.inl '' s) ∪ (sum.inr '' (set.univ : set β)) → s ⊕ β
+| ⟨sum.inl x, h⟩ := sum.inl ⟨x, begin
+      simp only [set.image_univ, set.mem_union_eq, set.mem_image, exists_eq_right, set.mem_range,
+        exists_false, or_false] at h,
+      exact h,
+    end⟩
+| ⟨sum.inr x, h⟩ := sum.inr x
+
+/-- Restricts a term on `α ⊕ fin n` to use only a set of the free variables indexed by `α`. -/
+def restrict_var_left {n : ℕ} (s : set α) (t : L.term (α ⊕ fin n))
+  (h : ∀ a, t.uses_var (sum.inl a) → a ∈ s) :
+  L.term (s ⊕ fin n) :=
```

Comment:
this seems easier to do with the generalization I suggested above. So that seems like the best way to go.